google/auth/impersonated_credentials.py

```diff
@@ -0,0 +1,190 @@
+
+    def __init__(self, root_credentials,  principal,
+                 new_scopes, delegates=None,
+                 lifetime=None):
+        """
+        Args:
+            root_credentials (google.auth.Credentials): The root credential
+                used as to acquire the impersonated credentials.
+            principal (str): The service account to impersonatge.
+            new_scopes (Sequence[str]): Scopes to request during the
+                authorization grant.
+            delegates (Sequence[str]): The chained list of delegates required
+                to grant the final access_token.
+            lifetime (int): Number of seconds the delegated credential should
+                be valid for (upto 3600).  If set, the credentials will
+                **not** get refreshed after expiration.  If not set, the
+                credentials will be refreshed every 3600s.
+        """
+
+        super(credentials.Credentials, self).__init__()
+
+        self._root_credentials = copy.copy(root_credentials)
+        self._root_credentials._scopes = _IAM_SCOPE
+        self._principal = principal
+        self._new_scopes = new_scopes
+        self._delegates = delegates
+        self._lifetime = lifetime
+        self.token = None
+        self.expiry = _helpers.utcnow()
+
+    @_helpers.copy_docstring(credentials.Credentials)
+    def refresh(self, request):
+        if (self.token is not None and self._lifetime is not None):
+            self.expiry = _helpers.utcnow()
+            raise exceptions.RefreshError(_LIFETIME_ERROR)
+        self._root_credentials.refresh(request)
+        self._updateToken(request)
+
+    @property
+    def expired(self):
+        return _helpers.utcnow() >= self.expiry
+
+    def _updateToken(self, req):
+        """Updates credentials with a new access_token representing
+        the impersonated account.
+
+        Args:
+            req (google.auth.transport.requests.Request): Request object to use
+                for refreshing credentials.
+
+        Raises:
+            TransportError: Raised if there is an underlying HTTP connection
+            Error
+            DefaultCredentialsError: Raised if the impersonated credentials
+            are not available.  Common reasons are
+            `iamcredentials.googleapis.com` is not enabled or the
+            `Service Account Token Creator` is not assigned
+        """
+
+        lifetime = self._lifetime
+        if (self._lifetime is None):
+            lifetime = _DEFAULT_TOKEN_LIFETIME_SECS
+        body = {
+            "delegates": self._delegates,
+            "scope": self._new_scopes,
+            "lifetime": str(lifetime) + "s"
+        }
+
+        iam_endpoint = _IAM_ENDPOINT.format(self._principal)
+        try:
+            headers = {
+                    'Content-Type': 'application/json',
+                    'Authorization': 'Bearer ' + self._root_credentials.token
+            }
+            response = req(url=iam_endpoint,
```

Comment:
Can you split this part into a separate private method, say `_make_iam_token_request`? It'll make the code more clear and easier to test.